sys/include/memarray.h

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2017 Tobias Heider <heidert@nm.ifi.lmu.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+/**
+ * @defgroup    sys_memarray memory array allocator
+ * @ingroup     sys
+ * @brief       memory array allocator
+ * @{
+ *
+ * @brief       pseudo dynamic allocation in static memory arrays
+ * @author      Tobias Heider <heidert@nm.ifi.lmu.de>
+ */
+
+#ifndef MEMARRAY_H
+#define MEMARRAY_H
+
+#include <stdint.h>
+#include <stdlib.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Memory pool
+ */
+typedef struct {
+    void *first_free;   /**< head of the free list */
+    size_t size;        /**< size of single list element */
+    size_t num;         /**< max number of elements in list */
+} memarray_t;
+
+/**
+ * @brief Define static memory pool
+ *
+ * @param[out] name       name for defined memarray pool
+ * @param[in]  structure  structure in single memory chunk
+ * @param[in]  numb       number of chunks in pool
+ */
+#define MEMARRAY(name, structure, numb)                                           \
```

Comment:
Please get rid of the macro.

Reply:
I agree that macros should be avoided, but I can not think of a better way to build an abstraction which declares the data array and memarray struct with a unique name and thus can be used to declare more than one such array. The only other option I see is letting the user do the declaration, which I think makes it more complicated to use. What do you think would be the best alternative?